Add proper constant terms for LinearCost and QuadraticCost. Add cost offsets to Gurobi and MOSEK solution costs.

EricCousineau-TRI opened the pull request:
This addresses Issue #3500, and adds constant terms for `LinearCost` and `QuadraticCost`, incorporates these terms in symbolic parsing, and ensures that the explicitly constant terms are added to the optimal cost if a solver does not support constant cost offsets.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6084)
<!-- Reviewable:end -->

EricCousineau-TRI commented:
+@russtedrake for feature review, please.
+@soonho-tri for platform review, please.

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-:-Kju23pa9UiWe56o-HgA:b2kcu55)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-:-KjuJyUdoawt4sN6taEH:b-3mz15d)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 13 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/cost.h, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCKd7GEzsAYEVnaXMf:-KkCKd7GEzsAYEVnaXMg:b-66yal2) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/cost.h#L89)):*
> ```objc
> 
>   // NOLINTNEXTLINE(runtime/explicit) This conversion is desirable.
>   LinearCost(const Eigen::Ref<const Eigen::VectorXd>& a, double b = 0.)
> ```

Maybe add the documentation here, saying that the default value for `b` is 0?

---

*[drake/solvers/cost.h, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCKpfDxRyAEJLYAmko:-KkCKpfDxRyAEJLYAmkp:b-bz1o5t) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/cost.h#L111)):*
> ```objc
>    * @param new_b New constant term.
>    */
>   void UpdateConstraint(const Eigen::Ref<const Eigen::RowVectorXd>& new_a,
> ```

Rename this function as `UpdateCoefficient`? This is not a constraint. And `UpdateConstraint` can mean the coefficient is unchanged, but the variables are changes, for example, changing `x(0) + 2 * x(1)` to `x(1) + 2 * x(0)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 2 of 13 files at r1, 2 of 7 files at r2.
Review status: 3 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/solvers/cost.h, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCLJ6hbMdtk6HbnaJc:-KkCLJ6ideoRbveOD_T-:b-145nc6) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/cost.h#L111)):*
> ```objc
>    * @param new_b New constant term.
>    */
>   void UpdateConstraint(const Eigen::Ref<const Eigen::RowVectorXd>& new_a,
> ```

`new_a` should be a `VectorXd` instead of `RowVectorXd`?

---

*[drake/solvers/cost.h, line 140 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCLe5OERNIvjpNKpae:-KkCLe5OERNIvjpNKpaf:b-tzwwym) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/cost.h#L140)):*
> ```objc
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(QuadraticCost)
> 
>   template <typename DerivedQ, typename Derivedb>
> ```

Add documentation for this function, that the default value of `c` is 0?

---

*[drake/solvers/create_cost.cc, line 48 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCNf7dUr0Qco1OU_6I:-KkCNf7dUr0Qco1OU_6J:bke31c6) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/create_cost.cc#L48)):*
> ```cpp
>       monomial_to_coeff_map, map_var_to_index, vars_vec.size(), &Q, &b,
>       &constant_term);
>   // Now add the quadratic constraint 0.5 * x' * Q * x + b' * x
> ```

Update the documentation?

---

*[drake/solvers/mathematical_program.h, line 2283 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCO36nbbnVW9Maz3jk:-KkCO36o3djU_wuofem8:b-5iba8w) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/mathematical_program.h#L2283)):*
> ```objc
>  * be captured by each respective solver.
>  */
> double ComputeExplicitlyConstantCost(const MathematicalProgram& prog);
> ```

So if we ask the nonlinear solver (like SNOPT) to solve a quadratic program, with constant term in the cost, what will happen?

Personally I prefer that each `xxx_solver.cc`, when looping through the cost, determine if it can add the constant term to the solver, or there should be a `constant_cost` stored inside the solver. Some of the solvers, such as nonlinear solvers,  Mosek, etc. (http://docs.mosek.com/8.0/capi/tutorial-lo-shared.html), support adding constant term to the cost; some other solvers, such as Gurobi, does not handle the constant term in the cost. So I believe this constant cost should live in each solver `gurobi_solver.cc`, `mosek_solver.cc`, etc, rather than in the `MathematicalProgram`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 1 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/solvers/cost.h, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCKd7GEzsAYEVnaXMf:-KkCO1IWDiIlvaPBY5DL:bpxnllo) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/cost.h#L89)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Maybe add the documentation here, saying that the default value for `b` is 0?
</blockquote></details>

Done. Rather than explicitly state the value, I've mentioned that it's optional, per the rationale here:
http://stackoverflow.com/a/14262130

---

*[drake/solvers/cost.h, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCKpfDxRyAEJLYAmko:-KkCOHhqRvbptZB71P6G:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/cost.h#L111)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Rename this function as `UpdateCoefficient`? This is not a constraint. And `UpdateConstraint` can mean the coefficient is unchanged, but the variables are changes, for example, changing `x(0) + 2 * x(1)` to `x(1) + 2 * x(0)`.
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCLJ6hbMdtk6HbnaJc:-KkCS6LjIBY0bzaLFKHf:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/cost.h#L111)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`new_a` should be a `VectorXd` instead of `RowVectorXd`?
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 140 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCLe5OERNIvjpNKpae:-KkCS8inAQKPLa4cpyIa:b-izpsi3) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/cost.h#L140)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Add documentation for this function, that the default value of `c` is 0?
</blockquote></details>

Done. Same rationale as above.

---

*[drake/solvers/create_cost.cc, line 48 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCNf7dUr0Qco1OU_6I:-KkCSIz8L2PnHLq4oUk7:b-z8bcf7) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/create_cost.cc#L48)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Update the documentation?
</blockquote></details>

Done. Removed the line, since it was no longer applicable and redundant.

---

*[drake/solvers/mathematical_program.h, line 2283 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkCO36nbbnVW9Maz3jk:-KkC_WcNWXc-Ct0uNItF:b-1xny6j) ([raw file](https://github.com/robotlocomotion/drake/blob/67cdc55049a517a90fc15e8f710c7298a341b3c6/drake/solvers/mathematical_program.h#L2283)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

So if we ask the nonlinear solver (like SNOPT) to solve a quadratic program, with constant term in the cost, what will happen?

Personally I prefer that each `xxx_solver.cc`, when looping through the cost, determine if it can add the constant term to the solver, or there should be a `constant_cost` stored inside the solver. Some of the solvers, such as nonlinear solvers,  Mosek, etc. (http://docs.mosek.com/8.0/capi/tutorial-lo-shared.html), support adding constant term to the cost; some other solvers, such as Gurobi, does not handle the constant term in the cost. So I believe this constant cost should live in each solver `gurobi_solver.cc`, `mosek_solver.cc`, etc, rather than in the `MathematicalProgram`.
</blockquote></details>

Done. Removed `ComputeExplicitlyConstantCost`, used `MSK_putcfix` for Mosek, and inlind constant term computation in Gurobi.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 13 files at r1, 7 of 12 files at r3.
Review status: 8 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/quadratic_program_examples.cc, line 350 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkChXzZyPguuMbHm_iX:-KkChXzZyPguuMbHm_iY:b-4irxfh) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/test/quadratic_program_examples.cc#L350)):*
> ```cpp
> }
> 
> void QuadraticProgram4::CheckSolution(SolverType solver_type) const {
> ```

Maybe add the check for the optimal cost here?

---

*[drake/solvers/test/quadratic_program_examples.cc, line 424 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkChe6l6Ebuv2iRfHkv:-KkChe6l6Ebuv2iRfHkw:b72o9ur) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/test/quadratic_program_examples.cc#L424)):*
> ```cpp
>     ASSERT_NO_THROW(RunSolver(&prog, solver));
> 
>     const auto& x_value = prog.GetSolution(x);
> ```

Add the check for the optimal cost?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 8 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/solvers/test/quadratic_program_examples.cc, line 350 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkChXzZyPguuMbHm_iX:-KkClO2LbQvIbVMwUGut:b-n1xotw) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/test/quadratic_program_examples.cc#L350)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Maybe add the check for the optimal cost here?
</blockquote></details>

Already done via `ExpectSolutionCostAccurate` - this will compute the cost from the bindings, and then compare that with `GetOptimalCost()` from the solver.
Was there another check that you felt was missing?

---

*[drake/solvers/test/quadratic_program_examples.cc, line 424 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkChe6l6Ebuv2iRfHkv:-KkClqiInxMujwMI1-RJ:b5qslni) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/test/quadratic_program_examples.cc#L424)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Add the check for the optimal cost?
</blockquote></details>

Already done, I believe - same as above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 12 files at r3.
Review status: 11 of 15 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-:-KkHoMQhg8z72u06wgPZ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 2 of 13 files at r1, 1 of 7 files at r2, 12 of 12 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/solvers/cost.h, line 118 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLmUJGP3QPX2CBZ-oX:-KkLmUJHvSX27fWq7jwj:b-xw13vo) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/cost.h#L118)):*
> ```objc
>                           double new_b = 0.) {
>     if (new_a.rows() != a_.rows()) {
>       throw std::runtime_error("Can't change the number of decision variables");
> ```

META: probably worth adding test case(s) for the cases throwing `runtime_error`? (here and below in `QuadraticCost` class).

---

*[drake/solvers/cost.h, line 134 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLYwxQQYsk4YJlTQq6:-KkLYwxQQYsk4YJlTQq7:b-41nclq) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/cost.h#L134)):*
> ```objc
>  private:
>   Eigen::VectorXd a_;
>   double b_;
> ```

BTW, nit: missing `{}` at the end.

---

*[drake/solvers/cost.h, line 145 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLZ-dCTvB1WdB05fXQ:-KkLZ-dCTvB1WdB05fXR:b-wqe1ov) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/cost.h#L145)):*
> ```objc
> 
>   /**
>    * Construct a cost of the form @f .5 x'Qx + b'x + c @f.
> ```

BTW, nit: `Construct` -> `Constructs`

---

*[drake/solvers/create_cost.cc, line 59 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLaGWLiQ19NEU5sHSX:-KkLaGWMZj4rNHDe1LQ5:b-3pkhlb) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/create_cost.cc#L59)):*
> ```cpp
>   double constant_term;
>   DecomposeLinearExpression(e, map_var_to_index, c, &constant_term);
>   return CreateBinding(make_shared<LinearCost>(c.transpose(), constant_term),
> ```

Just to confirm: Is `c` -> `c.transpose()` for fixing a bug?

---

*[drake/solvers/mathematical_program.h, line 720 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLbLxjOpYeyYOhPK7d:-KkLbLxjOpYeyYOhPK7e:btx3ljf) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/mathematical_program.h#L720)):*
> ```objc
>    * Adds a linear cost term of the form a'*x + b.
>    * @param e A linear symbolic expression.
>    * @pre{e is a linear expression a'*x + b, where each entry of x is a decision
> ```

BTW, I recently learned that `{`/`}` are not necessary (in `@pre{...}`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (pending discussion)

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-:-KkLnBeiGrofewsSQkQ5:b4saenz)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 6 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/solvers/cost.h, line 118 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLmUJGP3QPX2CBZ-oX:-KkMph9NP6qrpIJiXu_q:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/cost.h#L118)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

META: probably worth adding test case(s) for the cases throwing `runtime_error`? (here and below in `QuadraticCost` class).
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 134 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLYwxQQYsk4YJlTQq6:-KkMn3IHAWTkr7l2ORJL:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/cost.h#L134)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, nit: missing `{}` at the end.
</blockquote></details>

Done.

---

*[drake/solvers/cost.h, line 145 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLZ-dCTvB1WdB05fXQ:-KkMsePlXv9KDyNFBlgL:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/cost.h#L145)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, nit: `Construct` -> `Constructs`
</blockquote></details>

Done.

---

*[drake/solvers/create_cost.cc, line 59 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLaGWLiQ19NEU5sHSX:-KkMt4kwYbpa8xw1zXle:b-q0tp2q) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/create_cost.cc#L59)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Just to confirm: Is `c` -> `c.transpose()` for fixing a bug?
</blockquote></details>

Nope, it's to permit it to be used in `DecomposeLinearExpression()`, which requires that `c` be a `num_outputs x num_vars` matrix.

---

*[drake/solvers/mathematical_program.h, line 720 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLbLxjOpYeyYOhPK7d:-KkMu8mLZxcmgrq0RypB:bb46c8l) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/mathematical_program.h#L720)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, I recently learned that `{`/`}` are not necessary (in `@pre{...}`).
</blockquote></details>

Done. Thanks for the heads up!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
*[drake/solvers/create_cost.cc, line 59 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6084#-KkLaGWLiQ19NEU5sHSX:-KkMvL7ozydK4fxnWUBS:b-kualoh) ([raw file](https://github.com/robotlocomotion/drake/blob/d809e85e97785f5b167a7edbb00fd8c7dac94057/drake/solvers/create_cost.cc#L59)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Nope, it's to permit it to be used in `DecomposeLinearExpression()`, which requires that `c` be a `num_outputs x num_vars` matrix.
</blockquote></details>

I see. Never mind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 9 of 9 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6084)*
<!-- Sent from Reviewable.io -->
